openmc/mgxs/library.py

```diff
@@ -717,3 +725,307 @@ def load_from_file(filename='mgxs', directory='mgxs'):
+        if domain_names is not None:
+            cv.check_iterable_type('domain_names', filename, basestring)
+        if xs_ids is not None:
+            if isinstance(xs_ids, basestring):
+                # If we only have a string lets convert it now to a list
+                # of strings.
+                xs_ids = [xs_ids for i in range(len(self.domains))]
+            else:
+                cv.check_iterable_type('xs_ids', xs_ids, basestring)
+        else:
+            xs_ids = ['1g' for i in range(len(self.domains))]
+        cv.check_type('filename', filename, basestring)
+        cv.check_type('directory', directory, basestring)
+
+        # Construct the collection of the nuclides to report
+        if self.by_nuclide:
+            nuclides = self.all_mgxs[1][self.mgxs_types[-1]].get_all_nuclides()
+        else:
+            xs_type = 'macro'
+
+        # Make directory if it does not exist and build our filename
+        if not os.path.exists(directory):
+            os.makedirs(directory)
+        full_filename = os.path.join(directory, filename + '.xml')
+        full_filename = full_filename.replace(' ', '-')
+
+        # Initialize file
+        mgxs_file = openmc.MGXSLibrary(self.energy_groups)
+
+        # Set the scattering order as isotropic until
+        # support for higher orders are included in openmc.mgxs
+        order = 0
+
+        # Build XSdata objects
+        xsdatas = []
+
+        mat_names = {}
+        for i in range(len(self.domains)):
+
+            id = self.domains[i].id
+            if not self.by_nuclide:
+                # Build & add metadata to XSdata object
+                # (Use i here because k in nuclides will add chars to this)
+                if domain_names is None:
+                    name = 'set' + str(i + 1)
+                else:
+                    name = domain_names[i]
+                name += '.' + xs_ids[i]
+                xsdata = openmc.XSdata(name, self.energy_groups)
+                xsdata.order = order
+
+                mat_names[id] = name
+
+                # Now get xs data itself
+                if 'transport' in self.mgxs_types:
+                    if self.correction == 'P0':
+                        xsdata.set_total(self.all_mgxs[id]['transport'],
+                                         xs_type=xs_type, subdomains=(id,))
+                    else:
+                        msg = "The use of a transport cross section " + \
+                              "requires the correction attribute to be" + \
+                              "set to 'P0' to produce valid cross " + \
+                              "section libraries"
+                        raise ValueError(msg)
+                elif 'total' in self.mgxs_types:
+                    xsdata.set_total(self.all_mgxs[id]['total'],
+                                     xs_type=xs_type, subdomains=(id,))
+                if 'absorption' in self.mgxs_types:
```

Comment:
Good question.  I'll add a check near the start of the routine for types according to these rules:
- Fission is not required as a fixed source problem could be
the target.
- Absorption is required.
- A nu-scatter matrix is required.
  - Having both nu-scatter (@ whatever order) and scatter (isotropic)
    matrices is preferred
  - If only nu-scatter, need total (not transport), to
    be used in adjusting absorption
    (i.e., reduced_abs = tot - nuscatt)
  - If nu-scatter is anisotropic, self.correction must be None.
    (Not dealing with extended transport approximation yet)
- Either total or transport should be present.
  - Both can be available if one wants, but we should
    use whatever corresponds to self.correction (if P0: transport)

This is a little more rigorous than what is in the code you reviewed (which mostly only checked the scattering side), and also I've dropped the allowance in the code for scatter matrices to be used instead of nu-scatter.  I used to have it throw in a warning to allow for 'newb' mistakes, but I changed my mind.